File: Implement_Queue_Using_Stack.py
# Created by Aashish Adhikari at 12:02 PM 12/23/2020

class MyQueue(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """

        # Simulate two stacks using the standard deque from python
        from collections import deque
        self.IN_stack = deque() # to push in incoming order
        self.OUT_stack = deque() # to pop in FIFO order


    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: None
        """

        # Always push into the IN_stack
        self.IN_stack.append(x)


    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """



        _ = self.peek() # reusing the peek() to avoid redundant code

        return self.OUT_stack.pop()


    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        if len(self.OUT_stack) == 0:

            for idx in range(len(self.IN_stack)):
                self.OUT_stack.append(self.IN_stack.pop())

        return self.OUT_stack[len(self.OUT_stack)-1]

    # Since the questions says all peek operations are valid,
    # we do not have to handle the case when there is no element in the queue
    # and a peek() is requested.





    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        if len(self.IN_stack) == 0 and len(self.OUT_stack) == 0:
            return True
        return False




# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: Implement Queue using Stack.cpp
//Space Complexity- O(n)

class MyQueue {
public:
    /** Initialize your data structure here. */
    stack<int> st1;
    stack<int> st2;
    MyQueue() {
        
    }
    
    //Time Complexity- O(1)
    /** Push element x to the back of queue. */
    void push(int x) {
        st1.push(x);
    }
    
    //Time Complexity- O(n) worst case
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        while(!st1.empty()){
            st2.push(st1.top());
            st1.pop();
        }
        int ans=st2.top();
        st2.pop();
        while(!st2.empty()){
            st1.push(st2.top());
            st2.pop();
        }
        return ans;
    }
    
    //Time Complexity- O(n) worst case
    /** Get the front element. */
    int peek() {
        while(!st1.empty()){
            st2.push(st1.top());
            st1.pop();
        }
        int ans=st2.top();
        while(!st2.empty()){
            st1.push(st2.top());
            st2.pop();
        }
        return ans;
    }
    
    //Time Complexity- O(1)
    /** Returns whether the queue is empty. */
    bool empty() {
        return st1.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */